File: ios/LuminaLog/Core/Crypto/RecoveryCode.swift
import Foundation
import CryptoKit

/// Pure-crypto helper for the one-time **recovery code** backstop (spec §2).
///
/// The recovery code is a high-entropy secret shown to the user exactly once at
/// setup. It is never stored; instead a KEK is derived from it via HKDF-SHA256
/// and used to wrap the DEK (`WrappedKey`). If the iCloud Keychain KEK is ever
/// unavailable, the user re-enters the code, we re-derive the same KEK, and
/// unwrap the DEK.
///
/// Encoding is Crockford base32 (no ambiguous `I`, `L`, `O`, `U`) grouped in
/// fours (`XXXX-XXXX-…`). Normalization strips separators, upper-cases, and maps
/// the ambiguous glyphs a human might type back to their canonical digits.
enum RecoveryCode {

    /// Crockford base32 alphabet — 32 symbols, excludes I, L, O, U.
    static let alphabet = Array("0123456789ABCDEFGHJKMNPQRSTVWXYZ")

    /// Fixed application salt for the recovery-KEK derivation. Constant across
    /// all users/devices so the same code always derives the same KEK.
    static let hkdfSalt = Data("luminalog-recovery-kek-salt-v1".utf8)

    /// HKDF `info` (domain separation) for the recovery KEK.
    static let hkdfInfo = Data("luminalog-recovery-kek-v1".utf8)

    /// Bytes of entropy in a generated code (256-bit).
    static let entropyBytes = 32

    // MARK: - Generation

    /// Generate a fresh 256-bit recovery code in grouped Crockford base32.
    /// 256 bits → 52 base32 chars → 13 groups of 4 (`XXXX-XXXX-…`).
    static func generate() -> String {
        var bytes = [UInt8](repeating: 0, count: entropyBytes)
        let status = SecRandomCopyBytes(kSecRandomDefault, entropyBytes, &bytes)
        precondition(status == errSecSuccess, "SecRandomCopyBytes failed")
        return group(base32Encode(Data(bytes)))
    }

    // MARK: - KEK derivation

    /// Normalize a user-entered code: strip separators/whitespace, upper-case,
    /// and map ambiguous glyphs (`O`→`0`, `I`/`L`→`1`, `U`→`V`) to canonical
    /// symbols so a mistyped-but-unambiguous code still derives the right KEK.
    static func normalize(_ code: String) -> String {
        let stripped = code.uppercased().filter { $0 != "-" && !$0.isWhitespace }
        var mapped = ""
        mapped.reserveCapacity(stripped.count)
        for ch in stripped {
            switch ch {
            case "O": mapped.append("0")
            case "I", "L": mapped.append("1")
            case "U": mapped.append("V")
            default: mapped.append(ch)
            }
        }
        return mapped
    }

    /// Derive a 256-bit KEK from a recovery code via HKDF-SHA256 over the
    /// normalized code, with the fixed app salt and versioned info string.
    static func deriveKEK(from code: String) -> SymmetricKey {
        let ikm = SymmetricKey(data: Data(normalize(code).utf8))
        return HKDF<SHA256>.deriveKey(
            inputKeyMaterial: ikm,
            salt: hkdfSalt,
            info: hkdfInfo,
            outputByteCount: 32
        )
    }

    // MARK: - Wrap / unwrap

    /// Wrap `dek` under the KEK derived from `code`.
    static func wrap(dek: SymmetricKey, code: String) -> WrappedKey {
        WrappedKey.wrapping(dek: dek, under: deriveKEK(from: code))
    }

    /// Unwrap a recovery wrap using `code`. Fails closed on a wrong code.
    static func unwrap(_ wrap: WrappedKey, code: String) throws -> SymmetricKey {
        try wrap.unwrapping(under: deriveKEK(from: code))
    }

    // MARK: - Crockford base32

    /// Encode arbitrary bytes as an (ungrouped) Crockford base32 string.
    static func base32Encode(_ data: Data) -> String {
        var output = ""
        output.reserveCapacity((data.count * 8 + 4) / 5)
        var buffer = 0
        var bitsLeft = 0
        for byte in data {
            buffer = (buffer << 8) | Int(byte)
            bitsLeft += 8
            while bitsLeft >= 5 {
                let index = (buffer >> (bitsLeft - 5)) & 0x1F
                bitsLeft -= 5
                output.append(alphabet[index])
            }
        }
        if bitsLeft > 0 {
            let index = (buffer << (5 - bitsLeft)) & 0x1F
            output.append(alphabet[index])
        }
        return output
    }

    /// Insert a `-` every 4 characters for readability.
    static func group(_ s: String, size: Int = 4) -> String {
        var result = ""
        for (i, ch) in s.enumerated() {
            if i > 0 && i % size == 0 { result.append("-") }
            result.append(ch)
        }
        return result
    }
}

File: web/src/lib/crypto/keys/recoveryCode.ts
// Web port of iOS `Core/Crypto/RecoveryCode.swift`. Pure, no I/O.
//
// The recovery code is a high-entropy secret shown to the user exactly once at
// setup and stored nowhere. A KEK derived from it via HKDF-SHA256 wraps the DEK
// (see `../wrappedKey.ts`), which is the ONLY door a browser has: the other wrap
// slot is an iCloud Keychain key that no browser can reach.
//
// Every constant here is byte-exact with the Swift original. Changing any of
// them silently locks every existing user out of the web app, so they are
// pinned by the shared vector fixture in `vectors.test.ts`.

/** Crockford base32: 32 symbols, excluding the ambiguous I, L, O and U. */
export const ALPHABET = '0123456789ABCDEFGHJKMNPQRSTVWXYZ'

/** Fixed application salt, constant across all users and devices. */
const HKDF_SALT = new TextEncoder().encode('luminalog-recovery-kek-salt-v1')

/** HKDF `info`, for domain separation. */
const HKDF_INFO = new TextEncoder().encode('luminalog-recovery-kek-v1')

/** Bytes of entropy in a generated code (256-bit). */
const ENTROPY_BYTES = 32

/**
 * Normalize a user-entered code: strip separators and whitespace, upper-case,
 * and map the glyphs a human might type back to their canonical symbols, so a
 * mistyped-but-unambiguous code still derives the right KEK.
 */
export function normalize(code: string): string {
  let out = ''
  for (const ch of code.toUpperCase()) {
    if (ch === '-' || /\s/.test(ch)) continue
    if (ch === 'O') out += '0'
    else if (ch === 'I' || ch === 'L') out += '1'
    else if (ch === 'U') out += 'V'
    else out += ch
  }
  return out
}

/** Crockford base32, five bits at a time, most significant first. */
export function base32Encode(bytes: Uint8Array): string {
  let out = ''
  let buffer = 0
  let bitsLeft = 0
  for (const byte of bytes) {
    buffer = (buffer << 8) | byte
    bitsLeft += 8
    while (bitsLeft >= 5) {
      out += ALPHABET[(buffer >> (bitsLeft - 5)) & 0x1f]
      bitsLeft -= 5
    }
  }
  if (bitsLeft > 0) out += ALPHABET[(buffer << (5 - bitsLeft)) & 0x1f]
  return out
}

/** Insert a separator every `size` characters. */
export function group(s: string, size = 4): string {
  let out = ''
  for (let i = 0; i < s.length; i++) {
    if (i > 0 && i % size === 0) out += '-'
    out += s[i]
  }
  return out
}

/** Generate a fresh 256-bit recovery code: 52 base32 chars in 13 groups of 4. */
export function generate(): string {
  const bytes = crypto.getRandomValues(new Uint8Array(ENTROPY_BYTES))
  return group(base32Encode(bytes))
}

/**
 * Derive the 256-bit recovery KEK from `code` via HKDF-SHA256 over the
 * NORMALIZED code, with the fixed app salt and versioned info string.
 *
 * Returned non-extractable so the derived key can never leave the browser's key
 * store, and typed AES-GCM so it feeds `wrap`/`unwrap` directly.
 */
export async function deriveKEK(code: string): Promise<CryptoKey> {
  const ikm = new TextEncoder().encode(normalize(code))
  const ikmKey = await crypto.subtle.importKey('raw', ikm, 'HKDF', false, ['deriveBits'])
  const bits = await crypto.subtle.deriveBits(
    { name: 'HKDF', hash: 'SHA-256', salt: HKDF_SALT, info: HKDF_INFO },
    ikmKey,
    256,
  )
  return crypto.subtle.importKey('raw', bits, 'AES-GCM', false, ['encrypt', 'decrypt'])
}
